### PR TITLE
Add Fee Estimation Documentation & Best Practices

### LINKED ISSUE #153: Add best practices for fee estimation
Outline the available options when estimating fees for on-chain transactions that LDK wants to broadcast.

### PR DESCRIPTION
**Pull request by Beige-Coffee**
This PR is a first-draft submission for ```FeeEstimator``` documentation (including best practices). Once complete, it will close #153.

As of now, I submitted the documentation as a single file, since this will make it easier for others to read and provide comments. However, once the final edits are made, I think it would look best if I made it collapsable - like how the [Blockchain Data section](https://lightningdevkit.org/blockchain_data/introduction/) on the LDK Docs website has multiple sub-sections (ex: Introduction, Chain Activity, etc.). If this documentation followed that model, it would have the following sub-sections:

- Overview
- Fee Estimator
- Confirmation Target
- Best Practices
- Code Example

Once we finalize the edits and align on the right format, I will update the PR to include the necessary changes to the lightningdevkit.org website so that the fee estimation documentation will be available.

Looking forward to receiving feedback!

### CODE REVIEW

File: docs/fee_estimation.md [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+- Third-party fee estimates may affect the propagation of your pre-signed transactions at any time and, therefore, endanger the safety of channel funds.
+
+Running your own Bitcoin node can relieve some of these concerns, but it may require additional technical expertise and maintenance to operate properly. These trade-offs should be considered carefully when deciding on how your application will retrieve fee estimates.
+
+A savvy reader may have noticed that the above architecture is not very performant. LDK may generate a substantial number of fee-estimation calls in some cases. So, if we have to call an API every time we want a fee estimate, we'll add a significant amount of additional load to our application, and we may substantially slow HTLC handling. To ensure your application remains performant, you should pre-calculate and cache the fee estimates so that they can simply be retrieved when needed. Additionally, it's recommended to refresh fee estimates every 5-10 minutes so that your application has access to fresh freerates.
```
Comment (ConorOkus) [inline docs/fee_estimation.md:85]: We might also note that many of of these third parties are rate-limited so devs should handle that case to.

File: docs/fee_estimation.md [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+```
+pub fn get_fee_rate(&self, confirmation_target: ConfirmationTarget) -> u32 {
+    self.fee_rate_cache.get(&confirmation_target)
+        .map(|rate| cmp::max(*rate, FEERATE_FLOOR_SATS_PER_KW))
+        .expect("all ConfirmationTarget should be present in the cache")
+}
+```
+
+### Building A Fall-Back Plan 
+Great, we now have a function that will return a feerate, provided by mempool.space, for a given ```ConfirmationTarget```. But, there is a vulnerability here! What if mempool.space is down or our API call fails for some reason? If we're in urgent need of a feerate estimate, we'll need to return *something*. In a real production environment, we'd likely have a local mempool that we're sourcing feerates from and, perhaps, one ore more third-party services. However, for this demo, we'll hard-code fallback feerates and return those if we're unable to retrieve a feerate from mepool.space.
```
Comment (ConorOkus) [inline docs/fee_estimation.md:221]:
```suggestion
Great, we now have a function that will return a feerate, provided by mempool.space, for a given ```ConfirmationTarget```. But, there is a vulnerability here! What if mempool.space is down or our API call fails for some reason? If we're in urgent need of a feerate estimate, we'll need to return *something*. In a real production environment, we'd likely have a local mempool that we're sourcing feerates from and, perhaps, one ore more third-party services. However, for this demo, we'll hard-code fallback feerates and return those if we're unable to retrieve a feerate from mempool.space.
```
```suggestion
Great, we now have a function that will return a feerate, provided by mempool.space, for a given ```ConfirmationTarget```. But, there is a vulnerability here! What if mempool.space is down or our API call fails for some reason? If we're in urgent need of a feerate estimate, we'll need to return *something*. In a real production environment, we'd likely have a local mempool that we're sourcing feerates from and, perhaps, one ore more third-party services. However, for this demo, we'll hard-code fallback feerates and return those if we're unable to retrieve a feerate from mepool.space.
```
```suggestion
Great, we now have a function that will return a feerate, provided by mempool.space, for a given ```ConfirmationTarget```. But, there is a vulnerability here! What if mempool.space is down or our API call fails for some reason? If we're in urgent need of a feerate estimate, we'll need to return *something*. In a real production environment, we'd likely have a local mempool that we're sourcing feerates from and, perhaps, one ore more third-party services. However, for this demo, we'll hard-code fallback feerates and return those if we're unable to retrieve a feerate from mepool.space.
```
```suggestion
Great, we now have a function that will return a feerate, provided by mempool.space, for a given ```ConfirmationTarget```. But, there is a vulnerability here! What if mempool.space is down or our API call fails for some reason? If we're in urgent need of a feerate estimate, we'll need to return *something*. In a real production environment, we'd likely have a local mempool that we're sourcing feerates from and, perhaps, one ore more third-party services. However, for this demo, we'll hard-code fallback feerates and return those if we're unable to retrieve a feerate from mempool.space.
```
```suggestion
Great, we now have a function that will return a feerate, provided by mempool.space, for a given ```ConfirmationTarget```. But, there is a vulnerability here! What if mempool.space is down or our API call fails for some reason? If we're in urgent need of a feerate estimate, we'll need to return *something*. In a real production environment, we'd likely have a local mempool that we're sourcing feerates from and, perhaps, one ore more third-party services. However, for this demo, we'll hard-code fallback feerates and return those if we're unable to retrieve a feerate from mempool.space.
```

**Review by ConorOkus — APPROVED** [review]

File: docs/fee_estimation.md [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+	/// (ie 1 satoshi-per-byte rounded up to ensure later round-downs don't put us below 1 satoshi-per-byte).
+	///
+	/// The following unit conversions can be used to convert to sats/KW:
+	///  * satoshis-per-byte * 250
+	///  * satoshis-per-kbyte / 4
+	fn get_est_sat_per_1000_weight(&self, confirmation_target: ConfirmationTarget) -> u32;
+}
+```
+
+## The ```ConfirmationTarget```
```
[the remark this answers is not in the document]
Comment (Beige-Coffee) [reply docs/fee_estimation.md:37]: Noted! I removed the table and simply provided a link to the Rust API Docs instead.

File: docs/fee_estimation.md (hunk 1 above)
Comment (Beige-Coffee) [reply docs/fee_estimation.md:75]: There is some commentary about rate-limiting just above this section. Specifically, it says:

**- Third parties often rate-limit users such that they can only access a limited number of fee estimates within a given timeframe. If you require a certain service level agreement (SLA), it's recommended to run your own Bitcoin node or explore paid services with SLA guarantees, such as [mempool.space enterprise accounts](https://mempool.space/enterprise).**

Let me know if you think we should add more.

File: docs/fee_estimation.md (hunk 2 above)
Comment (Beige-Coffee) [reply docs/fee_estimation.md]: Good catch. Fixed!

**Comment by Beige-Coffee** [comment]
@jkczyz - I see you added yourself as a reviewer. Let me know if you have any feedback.

One question I'm still pondering is if this documentation should be a single page (ex: [Key Management](https://lightningdevkit.org/key_management/)) or if it should be collapsable (ex: [Blockchain Data section](https://lightningdevkit.org/blockchain_data/introduction/)). If it's collapsable, it would have the following sub-sections:

- Overview
- Fee Estimator
- Confirmation Target
- Best Practices
- Code Example

If you or @ConorOkus have an opinion, let me know. If not, I'll probably move forward with making it collapsable and adjusting this PR to include the associated code changes to the LDK website.

**Comment by jkczyz** [comment]
Sorry about the delay. Will take a look today.

Regarding the layout, I'm somewhat indifferent. Might even consider making the "Blockchain Data" section one page. I don't mind a long page as long everything is well-delineated and doesn't come across as too intimidating (e.g., large blocks of text). And if cross referencing other sections is useful, one page may be better suited.

File: docs/fee_estimation.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkczyz) [reply docs/fee_estimation.md:37]: Might be worth going through and cleaning up those docs at some point, FWIW.

**Review by jkczyz — COMMENTED** [review]
Really like the flow of the guide!

**Comment by Beige-Coffee** [comment]
Thanks for the detailed comments, @jkczyz!

I've updated the docs accordingly. The only feedback I haven't incorporated yet is creating a GitHub repo with an example code base for the reader to clone/fork. If we think that's important to include before merging, I can add it in.

**Pull request merged**